File: behaviorial/observer/python/concrete_subject.py
from observer import Observer
from subject import Subject, EventTypes


class EventManager(Subject):
    def __init__(self):
        self.observers: dict[str, set[Observer]] = {}

    def add_listener(self, event_name: EventTypes, observer: Observer):
        if event_name not in self.observers:
            self.observers[event_name] = set()
        self.observers[event_name].add(observer)

    def remove_listener(self, event_name: EventTypes, observer: Observer):
        if event_name in self.observers:
            self.observers[event_name].remove(observer)

    def notify(self, event_name: EventTypes, data):
        if event_name in self.observers:
            for observer in self.observers[event_name]:
                observer.update(data)

File: behaviorial/observer/typescript/concrete-subject.ts
import { Observer } from './observer';
import { EventTypes, Subject } from './subject';

export class EventManager implements Subject {
	observers: Map<EventTypes, Set<Observer>> = new Map();

	addListener(eventName: EventTypes, observer: Observer) {
		const observers = this.observers.get(eventName) || new Set();
		observers.add(observer);
		this.observers.set(eventName, observers);
	}

	removeListener(eventName: EventTypes, observer: Observer) {
		const observers = this.observers.get(eventName);
		if (observers) {
			observers.delete(observer);
			if (observers.size === 0) {
				this.observers.delete(eventName);
			}
		}
	}

	notify(eventName: EventTypes, data: any) {
		const observers = this.observers.get(eventName);
		if (observers) {
			observers.forEach(observer => observer.update(data));
		}
	}
}
